File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/service/money/MoneyConverter.java
package com.github.arhor.simple.expense.tracker.service.money;

import java.time.LocalDate;

import javax.money.CurrencyUnit;
import javax.money.MonetaryAmount;

public interface MoneyConverter {

    /**
     * Converts given monetary amount to the target currency using exchange rates on the given date.
     *
     * @param amount   amount to convert
     * @param currency target currency
     * @param date     conversion date
     * @param <T>      type of the concrete monetary amount
     *
     * @return converted monetary amount
     */
    <T extends MonetaryAmount> T convert(T amount, CurrencyUnit currency, LocalDate date);
}

File: app-server/src/main/kotlin/com/github/arhor/simple/expense/tracker/service/money/MoneyConverter.kt
package com.github.arhor.simple.expense.tracker.service.money

import java.time.LocalDate
import javax.money.CurrencyUnit
import javax.money.MonetaryAmount

interface MoneyConverter {

    /**
     * Converts given monetary amount to the target currency using exchange rates on the given date.
     *
     * @param amount   amount to convert
     * @param currency target currency
     * @param date     conversion date
     * @param T        type of the concrete monetary amount
     *
     * @return converted monetary amount
     */
    fun <T : MonetaryAmount> convert(amount: T, currency: CurrencyUnit, date: LocalDate): T
}